### DISCUSSION TITLE
One Edit Button für three textareas

### DISCUSSION THREAD

**Discussion by ckskr** [neuefische/questions#94] [Web]
Hello everybody, 

My problem is as follows: 
Integration of one Update-Button with wich I can edit and save all three lines of one daily diary entry.
Up till now I have an Update-Button which does weird things, but definitely not the ones I want. 

What did I try: 
A lot. Tried with Zustand - didn't work, deleted the whole branch. Now I am back to using UseState, but still stuck. Please help :). 

![image](https://user-images.githubusercontent.com/102232516/185913959-36f48b3d-cab1-42d6-bf2d-210ad6c9e2c8.png)

Here is my [github branch](https://github.com/ckskr/capstone-project/tree/user_story_2)

Here comes my code: 
```js
<StyledWrapper>
				{entries.map(entry => {
					return (
						<StyledCard key={entry.id}>
							<Date />
							<StyledH4>You were grateful for:</StyledH4>

							<ul>
								<StyledLi key={entry.id}>{entry.firstEntry}</StyledLi>
								<StyledLi key={entry.id}>{entry.secondEntry}</StyledLi>
								<StyledLi key={entry.id}>{entry.thirdEntry}</StyledLi>
								<div key={entry.id}>
									{entry.edit ? (
										<input
											type="text"
											value={entry.id}
											onChange={event => {
												setEntries(
													entries.map(newEntry => {
														return newEntry.name === entry.id
															? {
																	...newEntry,
																	name: event.target.value,
															  }
															: newEntry;
													})
												);
											}}
										/>
									) : (
										''
									)}

									<button
										type="button"
										onClick={() => {
											setEntries(
												entries.map(newEntry => {
													return newEntry.id === entry.id
														? {...newEntry, edit: !newEntry.edit}
														: newEntry;
												})
											);
										}}
									>
										{entry.edit ? 'save' : 'edit'}
									</button>
								</div>
							</ul>
						</StyledCard>
					);
				})}
			</StyledWrapper>
```

**Comment by CJAkkuec** [comment]
To edit an entry, you'd need to include a way for the entry's data to be used as the default values of the form you're using to create your entry data. Maybe I'm missing something, but it looks like you're only toggling the `edit` key of your entry's object (`{...newEntry, edit: !newEntry.edit}`), which is fine - but that's not automatically feeding the data into your form?

**Reply by ckskr** [reply]
I unfortunately don't have the level of coding skills to understand what you mean...

**Reply by CJAkkuec** [reply]
```
{entry.edit ? (
										<input
											type="text"
											value={entry.id}
											onChange={event => {
												setEntries(
													entries.map(newEntry => {
														
														return newEntry.name === entry.id
															? {
																	...newEntry,
																	name: event.target.value,
															  }
															: newEntry;
													})
												);
											}}
										/>
									) : (
										''
									)}
```

What happens here is that this input is rendered only when `entry.edit` is `true`, otherwise an empty string is rendered. The value of the input is always `entry.id`. I'm not sure changing the id is what you want to do? When this value is changed (which _can't_ be changed), this code is executed:
```
setEntries(
												entries.map(newEntry => {
													return newEntry.id === entry.id
														? {...newEntry, edit: !newEntry.edit}
														: newEntry;
												})
											);
```

You're mapping over your `entries`, and when `newEntry.name` is equal to `entry.id` (... you may want to compare newEntry.id to entry.id instead, by the way), you're returning an object that `newEntry` is spread into, with `name` having the value of the input field (that can't change).


I'm not sure you want any of this to happen, correct me if I'm wrong, though.

_What you may want to do_ is have an edit Button that will allow you to **re-use your form** with the input values being those of the entry you want to edit. You need a new state, though. Such as `postToEdit`, for example. So, when you click your edit Button, it will set the entry you want to edit as `postToEdit` and fill its values into your form. You should be able to edit the values and then have your save Button spread your edited entry into `entries`.

**Reply by gacsphere** [reply]
Hi @CJAkkuec,

I have a similar problem bzw. my problem starts where your explanation ends with the very last sentence. 
In my EditCardForm component, I can edit the values. From the edited values I render a new card and add it to the card-list. Actually I do it somehow like that: I render the edited card as new card and add it to my spreaded card-list. Additionally I delete the edit-card/originally edited card. Because if I wouldn't delete it, I would have the new created edited card, _PLUS_ the old card. But I'm not sure if this is the right way and I would prefer to have the edited card in my list on its original place. E.g. if I edit the _third_ card, I want to save it in the _third_ place in my list.
What I have now is that I can put it on top or at the end of my list. :(
How can I spread the edited card in the same place in the entries-list, where it was before? 

Here is an insight to my code:
https://github.com/gacsphere/capstone-project/pull/15

The appendCard function is in `index.js`. Later in the EditCardForm component, the appendCard function is called _onCreate_.
I edit the card in the EditCardForm component.

It would be nice, if you could help :)
Thanks Gesa

**Reply by ThomasF85** [reply]
First of all you need the information about the desired index.
Then you can add an element into the array at a specific index via splice for example:
```
    setCards((cards) => {
      const newCards = [...cards];
      newCards.splice(index, 0, {
        id: nanoid(),
        name,
        status
      });
      return newCards;
    });
```
